pkg/typesystem/typesystem_test.go

```diff
@@ -12,6 +12,490 @@ import (
+		//},
+		`this_has_no_entrypoints`: {
+			model: `
+			model
+				schema 1.1
+			type folder
+				relations
+					define parent: [folder#parent]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{false, false},
+		},
+		`computed_relation_has_entrypoint_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type document
+				relations
+					define editor: [user]
+					define viewer: editor`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`computed_relation_has_entrypoint_through_userset`: {
+			model: `
+			model
+			  schema 1.1
+			type user
+			type org
+			  relations
+				define member: [user]
+			type folder
+			  relations
+				define a2: [org#member]
+				define a1: a2`,
+			inputType:     "folder",
+			inputRelation: "a1",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`computed_relation_has_no_entrypoints_because_no_terminal_node`: {
+			model: `
+			model
+				schema 1.1
+			type folder
+				relations
+					define a2: a1
+					define a1: a2`,
+			inputType:     "folder",
+			inputRelation: "a1",
+			expectDetails: &RelationDetails{false, true},
+		},
+		`computed_relation_has_no_entrypoints`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			
+			type folder
+				relations
+					define parent: [document]
+					define viewer: editor from parent
+			
+			type document
+				relations
+					define parent: [folder]
+					define editor: viewer
+					define viewer: viewer from parent`,
+			inputType:     "folder",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{false, false},
+		},
+		`union_has_entrypoint_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			
+			type document
+				relations
+					define editor: [user]
+					define viewer: [document#viewer] or editor`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`union_has_no_entrypoint`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			
+			type document
+				relations
+					define editor: [document#viewer]
+					define viewer: [document#viewer] or editor`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{false, false},
+		},
+		`ttu_has_entrypoint_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type org
+				relations
+					define viewer: [user]
+			type folder
+				relations
+					define parent: [org]
+					define viewer: viewer from parent`,
+			inputType:     "folder",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`ttu_has_entrypoint_through_userset`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type org
+				relations
+					define viewer: [user]
+					define member: [user]
+			type folder
+				relations
+					define parent: [org#member]
```

Comment:
This is an invalid model. A tupleset relation cannot be directly related to a userset.

Reply:
i'm going to leave the test but document that `hasEntrypoints` assumes that all other model validations have run